webrender/doc/CLIPPING_AND_POSITIONING.md

```diff
@@ -0,0 +1,167 @@
+not affect rendering, by looking at the combined _outer rectangle_ and _inner
+rectangle_ of a `ClipChain` and the _outer rectangle_ and _inner rectangle_ of
+any `ClipNode` appended to the chain. An example of the goal of this process is
+to avoid having to render a mask for a large rounded rectangle when the rest of
+the clip chain constrains the content to an area completely inside that
+rectangle. Avoiding mask rasterization in this case and others has large
+performance impacts on WebRender.
+
+# Clipping and Positioning in the Display List
+
+Each non-structural WebRender display list item has
+ * A `ClipId` of a `SpatialNode` or `ClipNode` for positioning
+ * A `ClipId` of a `ClipNode` or a `ClipChain` for clipping
+ * An item-specific rectangular clip rectangle
+
+The positioning node determines how that item is positioned. It's assumed that
+the positioning node and the item are children of the same reference frame. The
+clipping node determines how that item is clipped. This should be fully
+independent of how the node is positioned and items can be clipped by any
+`ClipChain` regardless of the reference frame of their member clips. Finally,
+the item-specific clipping rectangle is applied directly to the item and should
+never result in the creation of a clip mask itself.
+
+Perhaps the most inconvenient holdover from the previous single-tree
+hierarchical design is that `SpatialNodes`, `ClipNodes`, and `ClipChains` all
+share a single `ClipId` id type. This means that the client must be a bit
+careful when using the API. For instance, when specifying the parent of
+`ClipNode` one can use the `ClipId` or another `ClipNode` or a `SpatialNode`,
+but not one for a `ClipChain`.
+
+WebRender's internal representation of clipping and positioning is not a perfect
+match to the display list representation of these concepts. This is due, again,
+to the evolutionary nature of the design. The general trend is that the display
+list gradually moves toward the internal representation. The most important of
+these incongruities is that while `ClipNodes`, sticky frames, and scroll frames
+are defined and simply return a ClipId, reference frames return a ClipId, but
+also are pushed and popped like stacking contexts.
+
+## Converting `ClipId` to global `ClipScrollTree` indices
+
+WebRender must access `ClipNodes` and `SpatialNodes` quite a bit when building
+scenes and frames, so it tries to convert `CiipIds`, which are already
+per-pipeline indices, to global scene-wide indices.  Internally this a
+conversion from `ClipId` into `SpatialNodeIndex`, `ClipNodeIndex` or
+`ClipChainIndex`. In order to make this conversion cheaper, the
+`DisplayListFlattner` assigns offsets for each pipeline and node type in the
+scene-wide `ClipScrollTree`.
+
+Nodes are added to their respective arrays sequentially as the display list is
+processed during "flattening." When encountering an iframe, the
```

Comment:
let's not include the "." in quote marks

Reply:
So weirdly this is the style in American English, which I admit is bizarre. Since the rest of the document is written in American English though, maybe it makes sense to be consistent here.